File: cpp/gof-cpp/visitor/Directory.cpp
#include "Entry.cpp"
#include "Visitor.cpp"
#include <string>
#include <vector>
class Directory : public Entry{
    private:
        std::string name;
        std::vector<Entry> directory;
    public :
        Directory(std::string n, int s){
            this->name = n;
        };
    public :
        std::string getName() override{
            return name;
        };
        int getSize() override{
            int size=0;
            for (auto it=directory.begin; it!=directory.end(); ++it){
                // std::cout << *it << std::endl;
                size+= it.getSize();
            };
            return size;
        };

    public:
        Entry add(Entry entry){
            directory.push_back(entry);
            return *this;
        };
    
    public:
        virtual void accept(Visitor& v) override
        {
            v.visit(*this);
        };
};

File: design-pattern/src/main/java/com/orcl/design/visitor/Directory.java
package com.orcl.design.visitor;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

// implements Iterable<Entry>を実装しない場合、以下起こる
// 1) iterable()メソッド Compile Error
// 2) 拡張For文への影響（Iterable Interfaceを実装しているため）

public class Directory extends Entry implements Iterable<Entry> {
    private String name;
    private List<Entry> directory = new ArrayList<>();

    public Directory(String name) {
        this.name = name;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public int getSize() {
        /*int size = 0;
        for (Entry entry: directory) {
            size += entry.getSize();
        }
        return size;*/
        SizeVisitor v = new SizeVisitor();
        accept(v);
        return v.getSize();
    }

    public Entry add(Entry entry) {
        directory.add(entry);
        return this;
    }

    @Override
    public Iterator<Entry> iterator() {
        return directory.iterator();
    }

    @Override
    public void accept(Visitor v) {
        v.visit(this);
    }
}
